contracts/src/access/enumerable.rs

```diff
@@ -0,0 +1,330 @@
+//! * All role management operations maintain consistency between AccessControl and enumeration state
+
+use alloc::{vec, vec::Vec};
+use alloy_primitives::{Address, FixedBytes, U256};
+use openzeppelin_stylus_proc::interface_id;
+pub use sol::*;
+use stylus_sdk::{
+    prelude::*,
+    storage::{StorageMap, StorageU256, StorageVec},
+};
+
+use super::control::{self, AccessControl, Error as AccessControlError, IAccessControl};
+use crate::utils::introspection::erc165::{Erc165, IErc165};
+
+#[cfg_attr(coverage_nightly, coverage(off))]
+mod sol {
+    use alloy_sol_macro::sol;
+
+    sol! {
+        /// Error returned when trying to query a role member at an invalid index.
+        #[derive(Debug)]
+        #[allow(missing_docs)]
+        error AccessControlEnumerableOutOfBounds(bytes32 role, uint256 index);
+    }
+}
+
+/// An error that occurred in the implementation of an [`AccessControlEnumerable`] contract.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// An error occurred in the base AccessControl contract.
+    AccessControl(AccessControlError),
+    /// Attempted to query a role member at an invalid index.
+    OutOfBounds(AccessControlEnumerableOutOfBounds),
+}
+
+/// Interface for the AccessControlEnumerable extension
+#[interface_id]
+pub trait IAccessControlEnumerable: IAccessControl {
+    /// Returns one of the accounts that have `role`.
+    /// `index` must be a value between 0 and {get_role_member_count}, non-inclusive.
+    fn get_role_member(role: FixedBytes<32>, index: U256) -> Result<Address, Error>;
+
+    /// Returns the number of accounts that have `role`.
+    fn get_role_member_count(role: FixedBytes<32>) -> U256;
+}
+
+/// State for role enumeration functionality.
+/// This struct maintains a mapping of roles to their member addresses,
+/// enabling efficient enumeration of role members.
+#[storage]
+#[derive(Debug)]
+pub struct RoleEnumeration {
+    /// Mapping from role to list of member addresses.
+    /// This tracks all accounts that have been granted each role for enumeration purposes.
+    /// The list is maintained in a deterministic order based on when members were added.
+    role_members: StorageMap<FixedBytes<32>, StorageVec<Address>>,
+}
+
+/// State of an enumerable access control contract.
+/// This struct combines the base AccessControl functionality with role enumeration capabilities,
+/// allowing for both role-based access control and enumeration of role members.
+///
+/// # Features
+///
+/// * All functionality from base AccessControl
+/// * Enumeration of role members
+/// * Tracking of role member count
+/// * Access to role members by index
+///
+/// # Note
+///
+/// The enumeration order of role members is deterministic but may not match the order
+/// in which roles were granted, especially after revocations.
+#[storage]
+#[derive(Debug)]
+pub struct AccessControlEnumerable {
+    /// Base access control functionality
+    #[borrow]
+    access: AccessControl,
```

Comment:
Hey @klingonaston!
Instead of having `AccessControl` as an attribute, please try to achieve the same design as we have for [Erc721 Enumerable extension](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/src/token/erc721/extensions/enumerable.rs) -- reference to `AccessControl` contract  should be passed to `AccessControlEnumerable` methods as you can see in [Erc721 Enumerable example](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/examples/erc721/src/lib.rs).